src/qiskit_qasm3_import/state.py

```diff
@@ -0,0 +1,205 @@
+    "crx": (_std.CRXGate, 1, 2),
+    "cry": (_std.CRYGate, 1, 2),
+    "crz": (_std.CRZGate, 1, 2),
+    "ch": (_std.CHGate, 0, 2),
+    "swap": (_std.SwapGate, 0, 2),
+    "ccx": (_std.CCXGate, 0, 3),
+    "cswap": (_std.CSwapGate, 0, 3),
+    "cu": (_std.CUGate, 4, 2),
+    "CX": (_std.CXGate, 0, 2),
+    "phase": (_std.PhaseGate, 1, 1),
+    "cphase": (_std.CPhaseGate, 1, 2),
+    "id": (
+        lambda: _std.UGate(0, 0, 0),  # Qiskit's "id" gate does not strictly have equal semantics.
+        0,
+        1,
+    ),
+    "u1": (_std.U1Gate, 1, 1),
+    "u2": (_std.U2Gate, 2, 1),
+    "u3": (_std.U3Gate, 3, 1),
+}
+
+_BUILTINS = {
+    "U": Symbol(
+        "U",
+        _std.UGate,
+        types.Gate([types.Angle() for _ in [None] * 3], [types.Qubit()]),
+        Scope.BUILTIN,
+    ),
+    "pi": Symbol("pi", math.pi, types.Float(const=True), Scope.BUILTIN),
+    "π": Symbol("π", math.pi, types.Float(const=True), Scope.BUILTIN),
+    "tau": Symbol("tau", math.tau, types.Float(const=True), Scope.BUILTIN),
+    "τ": Symbol("τ", math.tau, types.Float(const=True), Scope.BUILTIN),
+    "euler": Symbol("euler", math.e, types.Float(const=True), Scope.BUILTIN),
+    "ℇ": Symbol("ℇ", math.e, types.Float(const=True), Scope.BUILTIN),
+}
+
+
+class AddressingMode:
+    """Addressing mode for qubits in OpenQASM 3 programs.
+
+    This class is useful as long as we allow only physical or virtual addressing modes, but not
+    mixed. If the latter is supported in the future, this class will be modified or removed.
+
+    """
+
+    _Mode = enum.Enum("_Mode", ["UNKNOWN", "PHYSICAL", "VIRTUAL"])
+    (UNKNOWN, PHYSICAL, VIRTUAL) = (_Mode.UNKNOWN, _Mode.PHYSICAL, _Mode.VIRTUAL)
+
+    def __init__(self):
+        self._state = self.UNKNOWN
+
+    def set_physical_mode(self, node):
+        """Set the addressing mode to physical. On success return `True`, otherwise raise an
+        exception."""
+        if self._state == self.VIRTUAL:
+            raise_from_node(
+                node,
+                "Physical qubit referenced in virtual addressing mode. Mixing "
+                "modes not currently supported.",
+            )
+        self._state = self.PHYSICAL
+
+    def set_virtual_mode(self, node):
+        """Set the addressing mode to virtual.
+        On success return `True`, otherwise raise an exception."""
+        if self._state == self.PHYSICAL:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently "
+                "supported.",
+            )
+        self._state = self.VIRTUAL
+
+
+class SymbolTable:
+    __slots__ = ("symbols", "scope", "base", "global_symbols")
+
+    def __init__(self, scope: Scope, base: Optional["SymbolTable"] = None):
+        self.symbols = {}
+        self.scope = scope
+        self.base = base
+        self.global_symbols = base.global_symbols if base is not None else _BUILTINS.copy()
```

Comment:
I think it's better not to be accessing partial state of a `SymbolTable` (e.g. the backing dict for globals); it's likely to be a source of bugs if we need to do additional consistency management on `SymbolTable` operations, because the outer symbol table will end up being responsible for managing an secondary table's consistency, rather than just using the API to delegate.

If you're concerned about performance of fetching the `global_symbols`, then you could store a pointer to the global `SymbolTable` here, which is permitted to be the same pointer as `self.base`.

Reply:
I don't follow. Could you clarify, please? I took the basic idea from [here](https://github.com/Qiskit/qiskit-qasm3-import/pull/2#discussion_r1104892623). The current implementation is slightly awkward: the root symbol table creates a `dict` for globals. Then children copy a reference to this table. At least this encapsulates the global table. An alternative would be to make a reference to a `dict` in the global scope `_GLOBAL_SYMBOLS = _BUILTIS.copy()`. Then each table would use `_GLOBAL_SYMBOLS` directly for access.

Reply:
I'm saying that in this current design, `global_symbols` is a pointer to the `symbols` dict of a different `SymbolTable` instance, rather than being a pointer to the entire class.  That means that any mutations to that dict are going to be fragile for the `SymbolTable` that owns it, if the `SymbolTable` class evolves, because we're directly mutating internal state of some object without using the encapsulating methods, and mutating this state affects many other classes.

I'm suggesting that `global_symbols` should just be the "top" `SymbolTable` of the `SymbolTable.base` tree, and not just a part of the internal state of a `SymbolTable`.  Within that, the two things I suggested were either dropping `self.global_symbols` as an attribute and replace it with a method:
```python
class SymbolTable:
    def global_symbols(self) -> Self:
        table = self
        while table.base is not None:
            table = table.base
        return table
```
or if you think that it's critically important for performance to keep the attribute cached, then:
```python
class SymbolTable:
    def __init__(self, base: Optional[Self] = None):
        self.base = base
        if base is None:
            self.symbols = _BUILTINS.copy()
            self.global_symbols = self
        else:
            self.symbols = {}
            self.global_symbols = self.base.global_symbols
```

Reply:
I think I sorta follow what you mean.

This is not going to work:
```python
self.global_symbols = self
```
This `self.global_symbols = self.base.global_symbols` is similar to what I already had (not the same)--- a direct reference from leaves in the tree.

I think a good approach is to decide on an API that is insensitive to whether it's implemented more or less the way I already have it or something more like your idea. I'll try something....

Reply:
The symbol table is completely redesigned in 11c718b